### PR TITLE
Draft 03 $ref support

### PR DESCRIPTION
**Pull request by emmmile**
Some modifications on top of the work from @jamesaimonetti (mostly added local URI decoding and support for refs to array elements).

### CODE REVIEW

File: src/jesse_state.erl [hunk 1]
```diff
@@ -36,11 +37,17 @@
...
+        , find_schema/2
         ]).
 
 -export_type([ state/0
              ]).
 
+%% Includes
+-include("jesse_schema_validator.hrl").
+
+-define(schema_loader_fun, fun jesse_database:read/1).
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Invalid macro name schema_loader_fun on line 49. Use UPPER_CASE.

File: src/jesse_state.erl [hunk 2]
```diff
@@ -147,6 +162,20 @@ set_current_schema(State, NewSchema) ->
 set_error_list(State, ErrorList) ->
   State#state{error_list = ErrorList}.
 
+%% @doc Find a schema based on URI
+-spec find_schema(State :: state(), SchemaURI :: binary()) -> jesse:json_term() | ?not_found.
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 166 is too long: -spec find_schema(State :: state(), SchemaURI :: binary()) -> jesse:json_term() | ?not_found..

File: src/jesse_state.erl [hunk 3]
```diff
@@ -50,14 +57,12 @@
          , error_list         :: list()
          , error_handler      :: fun((#state{}) -> list() | no_return())
          , default_schema_ver :: atom()
+         , schema_loader_fun  :: fun((binary()) -> {ok, jesse:json_term()} | jesse:json_term() | ?not_found)
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 60 is too long:          , schema_loader_fun  :: fun((binary()) -> {ok, jesse:json_term()} | jesse:json_term() | ?not_found).

File: src/jesse_validator_draft3.erl [hunk 4]
```diff
@@ -202,6 +205,65 @@ check_value(Property, Value, Attrs, State) ->
...
+  Schema.
+
+%% @doc Decodes a $ref URI token. Replacen the ~0 and ~1 according to RFC 6901,
+%% run URI decode, then go back to binary.
+%% @private
+decode_path_element(Token) ->
+  String = binary:bin_to_list(Token),
+  String1 = http_uri:decode(String),
+  String2 = re:replace(String1, "\~0", "\~", [global,{return,list}]),
+  String3 = re:replace(String2, "\~1", "/", [global,{return,list}]),
```
Comment (gadgetci) [inline src/jesse_validator_draft3.erl]: According to **Elvis**:
> Missing space after "," on line 263

File: src/jesse_validator_draft3.erl [hunk 5]
```diff
@@ -202,6 +205,65 @@ check_value(Property, Value, Attrs, State) ->
...
+local_schema_array(Schema, []) ->
+  Schema.
+
+%% @doc Decodes a $ref URI token. Replacen the ~0 and ~1 according to RFC 6901,
+%% run URI decode, then go back to binary.
+%% @private
+decode_path_element(Token) ->
+  String = binary:bin_to_list(Token),
+  String1 = http_uri:decode(String),
+  String2 = re:replace(String1, "\~0", "\~", [global,{return,list}]),
```
Comment (gadgetci) [inline src/jesse_validator_draft3.erl]: According to **Elvis**:
> Missing space after "," on line 262

File: src/jesse_validator_draft3.erl [hunk 6]
```diff
@@ -202,6 +205,65 @@ check_value(Property, Value, Attrs, State) ->
...
+  Schema.
+
+%% @doc Decodes a $ref URI token. Replacen the ~0 and ~1 according to RFC 6901,
+%% run URI decode, then go back to binary.
+%% @private
+decode_path_element(Token) ->
+  String = binary:bin_to_list(Token),
+  String1 = http_uri:decode(String),
+  String2 = re:replace(String1, "~0", "\~", [global,{return,list}]),
+  String3 = re:replace(String2, "~1", "/", [global,{return,list}]),
```
Comment (gadgetci) [inline src/jesse_validator_draft3.erl]: According to **Elvis**:
> Missing space after "," on line 263

File: src/jesse_validator_draft3.erl [hunk 7]
```diff
@@ -202,6 +205,65 @@ check_value(Property, Value, Attrs, State) ->
...
+local_schema_array(Schema, []) ->
+  Schema.
+
+%% @doc Decodes a $ref URI token. Replacen the ~0 and ~1 according to RFC 6901,
+%% run URI decode, then go back to binary.
+%% @private
+decode_path_element(Token) ->
+  String = binary:bin_to_list(Token),
+  String1 = http_uri:decode(String),
+  String2 = re:replace(String1, "~0", "\~", [global,{return,list}]),
```
Comment (gadgetci) [inline src/jesse_validator_draft3.erl]: According to **Elvis**:
> Missing space after "," on line 262

**Comment by andreineculau** [comment]
oh wow :) now I have 3 PRs (this, #5 and #2) to look into around $ref

@emmmile @jamesaimonetti @emfa - could I kindly ask of you to provide some input on one another? without looking at all into the PRs, @emmmile or @emfa (both building on top of @jamesaimonetti) could get some inspiration from each other (completeness, correctness, etc), and update this PR or #5, so that I close the other two.

NOTE: @emfa's work in #5 builds up on #4 as well, so accepting #5 would maybe make things easier. I'm not hinting at anything, nor being biased, just stating the less obvious.

**Comment by emmmile** [comment]
@andreineculau actually you are right, sorry for having not checked #5 before.

#5 seems to cover all the cases, and it passes all the [tests](https://github.com/json-schema/JSON-Schema-Test-Suite), so I guess that is the important thing.

There is a lot of additional code for draft 04, but for what concerns draft 03 the work of @emfa looks good to me. So I would close this one, and work on #5.

**Comment by andreineculau** [comment]
thanks @emmmile!
closing in favour of #5

**Pull request closed without merging**